### DISCUSSION TITLE
Task is turning yaml into a programming language

### DISCUSSION THREAD

**Discussion by ghostsquad** [go-task/task#801] [General]
I feel like Task is taking the shape of YaaPL, Yaml as a Programming Language. It relies heavily on shell scripting, but via a go interpreter, meaning it will likely continue to be a bit buggy and/or miss specific features. There's a heavy reliance on text templating, but it seems for purposes that aren't strictly related to templating. Go text templating isn't even that easy to write or understand due to quoting, escaping, non-existent error handling, etc. 

It makes me question whether it's even the right approach to continue down this path. 

YAML isn't a programming language, it's a data serialization format. 

I feel like there will be a never ending stream of requests for features, ways to do things, complaints that specific functionality works one way, but they need it to work another way because folks won't have access to "raw primitives" the way they would in a typical programming language.

I'm starting to find something like [Mage](https://magefile.org/) to be a lot more enticing.

I'm curious what people thing about this?

**Comment by andreynering** [comment]
Different tools for different use cases.

If you want to write a script in a programming language Mage may be a good choice (although you could just use a bare `go run` as well).

Task is more about orchestrating stuff like building apps, which may require previous steps like code generation (where checksuming is important to skip when not necessary), etc.

In this case you'll mostly call external executables (Go, Docker, etc, etc). Doing that with a programming language may be inconvenient and require a lot more code, and you'll have to write some stuff like the mentioned checksuming of files yourself.

To be honest it may even make sense to use both: Task for the boring automation of calling stuff to build your app and Mage (or `go run`) when you need more flexibility and want to write the code yourself.

There's also the fact that Task is language agnostic. Many Task users are not Go programmers.

**Reply by ghostsquad** [reply]
You make some very good points. I think I've been trying to do too much with Task as of late, and seeing many limitations.

I think I understand why having "just yaml" is important, for orchestration, similar to GitHub actions without going too low level.

**Reply by fj** [reply]
@ghostsquad: You might also consider something like Pulumi if you're finding yourself needing more flexibility and power.

**Reply by ghostsquad** [reply]
> @ghostsquad: You might also consider something like Pulumi if you're finding yourself needing more flexibility and power.

I actually am using pulumi for infrastructure. But I still need a task runner (stateless). The other problem I run into is how the client is setup. Requiring specific tools installed, or even specific versions of tools, even stuff like `sed` and `awk` are different on Debian vs the GNU variants. Simple shell scripts are fine, but I often run into "wanting to do more", and shell scripting becomes untenable pretty quickly for anything more than a one-liner. Having to "call into" a CLI (such as one built with Go) requires me to then handle args/flags, stdout/err, and passing information back and forth between that program and others. It's actually the reason why I really like just keeping everything in a single robust language such as go.
